### PR TITLE
[tree widget] Add tooltips

### PR DESCRIPTION
**Pull request by chejennifer**
- add an info tooltip to explain the count and the greyed out stat var groups ([issue 1008](https://github.com/datacommonsorg/website/issues/1008))
- add tooltip for greyed out stat vars that gives list of placetypes and/or place names that have data for this stat var

dev updated with these changes

Screen Shot 2021-07-07 at 2 19 51 PM
Screen Shot 2021-07-07 at 2 19 58 PM

### CODE REVIEW

File: server/routes/api/stats.py [hunk 1]
```diff
@@ -224,3 +224,15 @@ def get_stats_all():
     return Response(json.dumps(dc.get_stats_all(dcids, stat_vars)),
                     200,
                     mimetype="application/json")
+
+
+@bp.route('/api/stats/stat-var-summary', methods=["POST"])
+@cache.cached(timeout=3600 * 24, query_string=True)
```
Comment (shifucun) [inline server/routes/api/stats.py:230]: can you check if the cache works for POST? I remember it has issue with memoize, but not sure for "query_string=True"

File: static/js/stat_var_hierarchy/stat_var_hierarchy.tsx [hunk 2]
```diff
@@ -283,4 +297,25 @@ export class StatVarHierarchy extends React.Component<
         return _.cloneDeep(resp.data["path"]).reverse();
       });
   }
+
+  private onMouseOverInfoIcon = () => {
+    const html =
+      "<ul><li>The number in parentheses represents the number of available" +
```
Comment (shifucun) [inline static/js/stat_var_hierarchy/stat_var_hierarchy.tsx:303]: need a space after "available"

File: static/js/stat_var_hierarchy/stat_var_hierarchy.tsx [hunk 3]
```diff
@@ -283,4 +297,25 @@ export class StatVarHierarchy extends React.Component<
...
+      "stat vars within the group that we have for the chosen place(s).</li>" +
```
Comment (shifucun) [inline static/js/stat_var_hierarchy/stat_var_hierarchy.tsx:304]: "have data for"?

File: static/js/stat_var_hierarchy/stat_var_hierarchy.tsx [hunk 4]
```diff
@@ -283,4 +297,25 @@ export class StatVarHierarchy extends React.Component<
...
+      "<li>Greyed out stat var groups have no available stat vars for the " +
+      "chosen place(s), but can still be expanded for you to explore.</li></ul>";
+    const topOffset = 30;
```
Comment (shifucun) [inline static/js/stat_var_hierarchy/stat_var_hierarchy.tsx:307]: moving the const to the top of the file?

File: static/js/stat_var_hierarchy/stat_var_section_input.tsx [hunk 5]
```diff
@@ -97,12 +106,53 @@ export class StatVarSectionInput extends React.Component<
...
+          onMouseOut={() => hideTooltip()}
         >
           {this.props.statVar.displayName}
         </label>
       </form>
     );
   }
+
+  private mouseMoveAction = (e) => {
+    let html = "loading stat var summary...";
```
Comment (shifucun) [inline static/js/stat_var_hierarchy/stat_var_section_input.tsx:121]: how long does it load?

File: static/js/stat_var_hierarchy/stat_var_section_input.tsx [hunk 6]
```diff
@@ -97,12 +106,53 @@ export class StatVarSectionInput extends React.Component<
...
+      }
+      html =
+        availablePlaceTypes.length > 0
+          ? "This stat var has no data for any of the chosen places." +
+            "You can try these types of places instead. <ul>"
+          : "Sorry, this stat var is not supported by this tool.";
+      for (const placeType of availablePlaceTypes) {
+        const placeSummary = this.props.summary.placeTypeSummary[placeType];
+        const placeList = placeSummary.topPlaces.map((place) => place.name);
+        html +=
```
Comment (shifucun) [inline static/js/stat_var_hierarchy/stat_var_section_input.tsx:140]: do you think it works to make the place names hyper link? Then user can directly go to the timeline page of that place.

File: static/js/stat_var_hierarchy/stat_var_section_input.tsx [hunk 7]
```diff
@@ -97,12 +106,53 @@ export class StatVarSectionInput extends React.Component<
...
+          this.context.statVarHierarchyType === StatVarHierarchyType.TIMELINE
+            ? `<li>${placeType} (eg. ${placeList.join(", ")})</li>`
+            : `<li>${placeType}</li>`;
+      }
+      html += "</ul>";
+    }
+    const left = e.pageX;
+    const topOffset = 10;
```
Comment (shifucun) [inline static/js/stat_var_hierarchy/stat_var_section_input.tsx:148]: nit: put constant at top of the file

File: server/routes/api/stats.py (hunk 1 above)
Comment (chejennifer) [reply server/routes/api/stats.py:230]: yes, you're right, the cache has problems. I removed the caching.

Comment (chejennifer) [reply server/routes/api/stats.py]: you're right, cache has issues, so I removed the caching.

File: static/js/stat_var_hierarchy/stat_var_hierarchy.tsx (hunk 2 above)
Comment (chejennifer) [reply static/js/stat_var_hierarchy/stat_var_hierarchy.tsx]: done

File: static/js/stat_var_hierarchy/stat_var_hierarchy.tsx (hunk 3 above)
Comment (chejennifer) [reply static/js/stat_var_hierarchy/stat_var_hierarchy.tsx]: done

File: static/js/stat_var_hierarchy/stat_var_hierarchy.tsx (hunk 4 above)
Comment (chejennifer) [reply static/js/stat_var_hierarchy/stat_var_hierarchy.tsx]: done

File: static/js/stat_var_hierarchy/stat_var_section_input.tsx (hunk 5 above)
Comment (chejennifer) [reply static/js/stat_var_hierarchy/stat_var_section_input.tsx:124]: The longest I've seen is 300ms

File: static/js/stat_var_hierarchy/stat_var_section_input.tsx (hunk 6 above)
Comment (chejennifer) [reply static/js/stat_var_hierarchy/stat_var_section_input.tsx:143]: I think we've discussed how links don't really belong in tooltips..

File: static/js/stat_var_hierarchy/stat_var_section_input.tsx (hunk 7 above)
Comment (chejennifer) [reply static/js/stat_var_hierarchy/stat_var_section_input.tsx]: done

File: static/js/stat_var_hierarchy/stat_var_section_input.tsx [hunk 8]
```diff
@@ -97,12 +109,51 @@ export class StatVarSectionInput extends React.Component<
...
+          (placeType) =>
+            placeType in USA_CHILD_PLACE_TYPES && placeType !== "Country"
+        );
+      }
+      html =
+        availablePlaceTypes.length > 0
+          ? "This stat var has no data for any of the chosen places." +
+            "You can try these types of places instead. <ul>"
+          : "Sorry, this stat var is not supported by this tool.";
+      for (const placeType of availablePlaceTypes) {
```
Comment (shifucun) [inline static/js/stat_var_hierarchy/stat_var_section_input.tsx:140]: another thing is to the order of place types, can defer to later PR

**Review by shifucun — APPROVED** [review]

File: static/js/stat_var_hierarchy/stat_var_section_input.tsx (hunk 6 above)
Comment (chejennifer) [reply static/js/stat_var_hierarchy/stat_var_section_input.tsx:143]: I think we've discussed that tooltips aren't really supposed to be used for the user to take actions? It should be more of a simple informational kind of thing ..

File: static/js/stat_var_hierarchy/stat_var_section_input.tsx (hunk 8 above)
Comment (chejennifer) [reply static/js/stat_var_hierarchy/stat_var_section_input.tsx:140]: Ok, what would be the order of place types?

File: static/js/stat_var_hierarchy/stat_var_section_input.tsx (hunk 6 above)
Comment (shifucun) [reply static/js/stat_var_hierarchy/stat_var_section_input.tsx:143]: I can see that, but from usability point of view, navigating to the hint places take some more effort (search place, then find the stat vars). We can gather thoughts from the team too, can update in later PR if necessary.

Comment (chejennifer) [reply static/js/stat_var_hierarchy/stat_var_section_input.tsx:143]: Yeah, I see what you mean, but in that case, maybe tooltip isn't the best way to show this information. Also trying to give users examples to navigate to will be a little more complicated for scatter and map tools where we need to know the parent place.

But yes, let's gather thoughts from the team!

**Pull request merged**